File: src/services/Services.NestJS/tool-registry.ts
import { readFileSync } from 'fs';

export interface ToolDefinition {
  name: string;
  description: string;
  script?: string;
  wrapper?: string;
  keywords?: string[];
  parameters?: string[];
  examples?: string[];
  result_example?: unknown;
  expected_json_output?: unknown;
  output_description?: string;
  aliases?: string[];
}

interface ToolRegistryDocument {
  tools?: ToolDefinition[];
}

export class ToolRegistry {
  private readonly tools: ToolDefinition[];

  constructor(registryPath: string) {
    const registry = JSON.parse(readFileSync(registryPath, 'utf8')) as ToolRegistryDocument;

    if (!Array.isArray(registry.tools)) {
      throw new Error('Tool registry must contain a tools array');
    }

    this.tools = registry.tools;
  }

  getTools(): ToolDefinition[] {
    return this.tools.map((tool) => ({ ...tool }));
  }

  findTool(toolName: string): ToolDefinition {
    const tool = this.tools.find(
      (item) => item.name === toolName || (item.aliases ?? []).includes(toolName)
    );

    if (!tool) {
      throw new Error(`Unknown tool: ${toolName}`);
    }

    return { ...tool };
  }
}

File: src/services/Services.DotNet/ToolRegistry.cs
using System.Text.Json;

namespace Services.DotNet;

public sealed class ToolRegistry : IToolRegistry
{
    private readonly IReadOnlyList<ToolDefinition> _tools;

    public ToolRegistry(string registryPath)
    {
        if (string.IsNullOrWhiteSpace(registryPath))
        {
            throw new ArgumentException("Registry path is required", nameof(registryPath));
        }

        _tools = LoadTools(registryPath);
    }

    public IReadOnlyList<ToolDefinition> GetTools()
    {
        return _tools;
    }

    public ToolDefinition FindTool(string toolName)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(toolName);

        var tool = _tools.FirstOrDefault(item =>
            string.Equals(item.Name, toolName, StringComparison.OrdinalIgnoreCase) ||
            item.Aliases.Any(alias => string.Equals(alias, toolName, StringComparison.OrdinalIgnoreCase)));

        return tool ?? throw new ArgumentException($"Unknown tool: {toolName}", nameof(toolName));
    }

    private static IReadOnlyList<ToolDefinition> LoadTools(string registryPath)
    {
        using var stream = File.OpenRead(registryPath);
        var document = JsonSerializer.Deserialize<ToolRegistryDocument>(stream);

        if (document?.Tools is not { Count: > 0 })
        {
            throw new InvalidOperationException("Tool registry must contain a tools array");
        }

        return document.Tools;
    }

    private sealed class ToolRegistryDocument
    {
        [System.Text.Json.Serialization.JsonPropertyName("tools")]
        public List<ToolDefinition> Tools { get; init; } = [];
    }
}